db/compaction/compaction_job.cc

```diff
@@ -396,29 +534,43 @@ void CompactionJob::GenSubcompactionBoundaries() {
           *(c->mutable_cf_options()), out_lvl,
           c->immutable_options()->compaction_style, base_level,
           c->immutable_options()->level_compaction_dynamic_level_bytes)));
-  uint64_t subcompactions =
-      std::min({static_cast<uint64_t>(ranges.size()),
-                static_cast<uint64_t>(c->max_subcompactions()),
-                max_output_files});
-
-  if (subcompactions > 1) {
-    double mean = sum * 1.0 / subcompactions;
-    // Greedily add ranges to the subcompaction until the sum of the ranges'
-    // sizes becomes >= the expected mean size of a subcompaction
-    sum = 0;
-    for (size_t i = 0; i + 1 < ranges.size(); i++) {
-      sum += ranges[i].size;
-      if (subcompactions == 1) {
-        // If there's only one left to schedule then it goes to the end so no
-        // need to put an end boundary
-        continue;
-      }
-      if (sum >= mean) {
-        boundaries_.emplace_back(ExtractUserKey(ranges[i].range.limit));
-        subcompactions--;
-        sum = 0;
-      }
+
+  // Get the number of planned subcompactions, may update reserve threads
+  // and update extra_num_subcompaction_threads_reserved_
+  uint64_t num_planned_subcompactions = InitializeNumPlannedSubcompactions(
+      max_output_files, static_cast<uint64_t>(ranges.size()));
+
+  TEST_SYNC_POINT_CALLBACK("CompactionJob::GenSubcompactionBoundaries:0",
+                           &num_planned_subcompactions);
+  if (num_planned_subcompactions == 1) return;
+
+  double mean = sum * 1.0 / num_planned_subcompactions;
+  // Greedily add ranges to the subcompaction until the sum of the ranges'
+  // sizes becomes >= the expected mean size of a subcompaction
+  sum = 0;
+  uint64_t num_actual_subcompactions = 0U;
+  for (size_t i = 0; i + 1 < ranges.size(); i++) {
+    sum += ranges[i].size;
+    if (num_actual_subcompactions + 1 == num_planned_subcompactions) {
+      num_actual_subcompactions++;
+      // If there's only one left to schedule then it goes to the end so no
+      // need to put an end boundary
+      continue;
```

Comment:
The condition (`num_actual_subcompactions + 1 == num_planned_subcompactions`) will no longer hold upon continuing to the next iteration. Can we init `num_actual_subcompactions` to 1, change the condition to `num_actual_subcompactions == num_planned_subcompactions`, and remove the increment on L555?

Reply:
sure! Will do.